fastfield_codecs/src/null_index/sparse.rs

```diff
@@ -0,0 +1,535 @@
+use std::io::{self, Write};
+
+use ownedbytes::OwnedBytes;
+
+/// `SparseCodec` is the codec for data, when only few documents have values.
+/// In contrast to `DenseCodec` opening a `SparseCodec` causes runtime data to be produced, for
+/// faster access.
+///
+/// The lower 16 bits of doc ids are stored as u16 while the upper 16 bits are given by the block
+/// id.
+///
+/// # Serialized Data Layout
+/// The data starts with the raw data of u16 encoded positions of the blocks. [u16 LE, .. #repeat*;
+/// Desc: Positions with values in a block]
+///
+/// The data is followed by block metadata, to know which area of the raw data belongs to which
+/// block. Only metadata for blocks with elements is recorded to
+/// keep the overhead low for scenarios with many very sparse columns. The block metadata consists
+/// of the block index and the number of values in the block.
+/// The last u16 is storing the number of metadata blocks.
+/// [u16 LE, .. #repeat*; Desc: Positions with values in a block][(u16 LE, u16 LE), .. #repeat*;
+/// Desc: (Block Id u16, Num Elements u16)][u16 LE; Desc: num blocks with values u16]
+///
+/// # Opening
+/// When opening the data layout, the data is expanded to `Vec<SparseCodecBlock>`, where the
+/// index is the block index.
+pub struct SparseCodec {
+    data: OwnedBytes,
+    blocks: Vec<SparseCodecBlock>,
+}
+
+/// A block consists of max u16 values
+#[derive(Debug)]
+struct SparseCodecBlock {
+    /// The number of values in the block
+    num_vals: u16,
+    /// The number of values set before the block
+    offset: u32,
+}
+
+impl SparseCodecBlock {
+    fn empty_block(offset: u32) -> Self {
+        Self {
+            num_vals: 0,
+            offset,
+        }
+    }
+
+    fn new(num_vals: u16, offset: u32) -> Self {
+        Self { num_vals, offset }
+    }
+    #[inline]
+    fn value_at_idx(&self, data: &[u8], idx: u16) -> u16 {
+        let data_start_index = (self.offset + idx as u32) * 2;
+        u16::from_le_bytes([
+            data[data_start_index as usize],
+            data[data_start_index as usize + 1],
+        ])
```

Comment:
That's a bummer, but the compiler sucks at that...
It does two boundary checks and two loads.

```suggestion
        let start_offset: usize = (self.offset + idx as u32) * 2 as usize;
        let bytes: [u8; 2] = data[start_offset..start_offset+2].try_into().unwrap();
        u16::from_le_bytes(bytes)
```
generates much nicer assembly